### DISCUSSION TITLE
Can I selectively display glyphs only if we're on a specific input?

### DISCUSSION THREAD

**Discussion by TheEmbracedOne** [Orange-Panda/Rewired-Glyphs#10] [Q&A] [answered]
I'm displaying glyphs in context (ie "[B] Exit" - imagine for a moment that that is an xbox "B" glyph there), and while I display prompts for mosue and keyboard in my game's "controls" setting, I do not want to display glyphs for mouse/keyboard in context.

So if I have a text that says "<Glyph Cancel> Exit", I want it to display "[B] Exit" when on gamepad, but I want it to display an empty (transparent) glyph when on mouse/keyboard, resulting in "  Exit" (I already put in a null/empty/uninitialised glyph sprite on the Input Glyph Collection but it only shows up when there is no keybinding on that key at all.

I tried "<Glyph Cancel type="joystick"> Exit", hoping it'd only show the joystick icon if the input is joystick, but unfortunately it just force-shows [B] even if I dont have a controller connected at all.

Is what I want to achieve here possible at all? Or should I look into a custom solution where I have maybe two texts in the same place, one with "<Glyph Cancel> Exit" to show when the input is joystick, and a simple "Exit" when it's mouse/keyboard? Thank you.

**Comment by Orange-Panda** [comment]
Hmm, I don't believe this is possible with the native components currently. It can definitely be done with a custom component (by querying Rewired player for recent controller the same way InputGlyphs does) but maybe this is a considerable feature for the built in components. 

Which component is this feature desired for? Is your use case hiding keyboard/mouse glyphs since these are clickable buttons?

**Comment by TheEmbracedOne** [comment]
I'm using the GlyphRichTextFormatter, and that is exactly my use case: "hiding keyboard/mouse glyphs since these are clickable buttons"!

**Comment by Orange-Panda** [comment]
This does seem like a really useful addition so I will look into adding this functionality natively into the GlyphDisplay components. I will get back to you with any updates I have for the feature.

**Comment by TheEmbracedOne** [comment]
Thank you, I await news patiently! Thank you for your continued, amazing work.

**Reply by Orange-Panda** [reply]
I've implemented this feature on the development branch. It will be included in the next release which will happen after I have done some more testing.

**Comment by Orange-Panda** [comment, marked as answer]
This feature has been published with [2.0.0](https://github.com/Orange-Panda/Rewired-Glyphs/discussions/13). It can be utilized in glyph rich text with the `hideKBM` parameter or by toggling `Hide Keyboard Mouse Glyphs` on other InputDisplay components.

**Reply by TheEmbracedOne** [reply]
Thank you so much!! I'm grabbing it right now.

**Reply by TheEmbracedOne** [reply]
Just wanted to follow up and say this works nicely and through some experimentation I was able to get the behaviour I wanted. I'll just leave a question here in case it is related to my setup:
I am running into an issue where some glyphs seem to display the wrong icon. It worked before, which is why I'm surprised it no longer does, and I remember this being an issue before 2.0.0, I just figured once I regenerate the sprite sheets it'll fix itself. (For reference I re-set up the Glyph Maps and Glyph Collections after importing 2.0.0)

For example, `<Glyph Cancel set=nobg>` should display Circle on a playstation controller but it displays R2. I quadruple checked, and my setup seems to be right:

1. My Action "Cancel" is hooked up to Circle:
![Unity_6sMferPtCZ](https://github.com/Orange-Panda/Rewired-Glyphs/assets/16197906/e7617f21-7d96-41f7-b92a-1a11621279b0)

2. And the correct glyph hooked up for Circle:
![Unity_IJuurNg7On](https://github.com/Orange-Panda/Rewired-Glyphs/assets/16197906/4060d67b-9ea3-4883-93fb-8426782ebad9)

3. I have tried deleting all the Resources\Sprite Assets and regenerated the spritesheets, but the issue persists.

but in the end, this is what variations of `<Glyph Cancel>` display as ingame:
![Unity_xNCGyu87mq](https://github.com/Orange-Panda/Rewired-Glyphs/assets/16197906/71226e07-8e69-4351-9f95-6bb963c6606c)

 What could be going wrong?

**Reply by Orange-Panda** [reply]
May I see where this glyph map is entered into your `GlyphCollection`? 

Some considerations with the information so far:
- I have a suspicion that the map is defined on the collection as a different type than it is designed for.
- Is this supposed to be a `Template Map` or a `Controller Map`?
    - If this is defined under `Template Map` on the glyph collection do the following: on the glyph map assign a `Controller Data Files` asset, press `Set by Template -> Gamepad`.
    - If this is defined under `Controller Map` on the glyph collection do the following: on the glyph map assign a `Controller Data Files` asset, press `Set by Template -> {Sony ____}` (depending on this device).
    - This will validate the actions to check if they have the correct name for their input id (the editable int field)

Also worth noting that it doesn't make sense to have a glyph map in both the template and controller map on the collection so you may have to make two glyph maps: one glyph map for the device elements and one for the template elements.

**Reply by TheEmbracedOne** [reply]
Thank you for the clarification!

This Glyph Map is present in the Glyph Collection as both a Template and Controller Map. So what I did now, I removed the xbox and playstation Glyph Maps from Controller Maps and left them as Template Maps (I dont display joystick glyphs based on actual input device, it's a user setting).

I have had Controller Data Files assigned on all my Glyph Maps, I unassigned and re-assign them now to be safe. I also set the "Set by Template -> Gamepad", and now I see some warnings and errors:
![image](https://github.com/Orange-Panda/Rewired-Glyphs/assets/16197906/a1e8e8e5-972b-4189-9d65-a7f3276687a6)

I'll do the required actions to get these errors/warnings. I'll probably just generate defaults and fill out the Glyph Maps again, that should eliminate these issues I think, and from what I understand it'll properly unify the terms as well (so instead of Playstation's Circle and Xbox's B, I'll just have "Action Bottom Row 2" which is clearer for my use case).

I'm sure doing this will elimiate the problem, but I'll report back the results once I set this up! Thank you again.

**Reply by Orange-Panda** [reply]
You're welcome! 

Regarding the "Action Row" description it would be valid to ignore the warning and rename it to match the symbol (Cross for playstation bottom 1 for example) but that's up to how you'd prefer to present the descriptions in your project, which it sounds like the generic term may be your preference

Let me know if you encounter any other issues or questions

**Reply by TheEmbracedOne** [reply]
Just came back to report that doing the above fixed my issue. I just regenerated the xbox and playstation maps and filled it out (quite alright with the "Action Row" descriptions by now. If Rewired calls them that I find it more cohesive that the maps call it that way). All works perfectly now. Thank you so much!
